### DISCUSSION TITLE
How to get a project's version from meson.build file?

### DISCUSSION THREAD

**Discussion by rffontenelle** [mesonbuild/meson#10154] [Q&A] [answered]
I have this project's meson.build file and I would like to obtain it's version in programmatic way.

Talking about shell scripting, I could do `grep ' version:' meson.build | cut -d\' -f2` when `version: x.x.x` is not in the same line as the `project()` function. If in the same, would be little be longer.

Is there a more Meson-way for me to extract this version from meson.build?

**Comment by eli-schwartz** [comment]
The JSON introspection data has this information, but only once you generate the builddir.

The `meson introspect` tool has *undocumented* functionality in that you can pass a file argument pointing to the root meson.build instead of a directory argument pointing to the root builddir. Meson will then do a full AST parse (infinite recursion IIRC, very slow and all that) to do a best effort static analysis of the build files and extract as much information as it can. That can be used to (eventually) query the first line of the root meson.build and extract the projectinfo. :)

**Comment by xclaesse** [comment, marked as answer]
`meson introspect meson.build --projectinfo` gives you a json that contains to version. Maybe we should actually document that feature.

**Reply by rffontenelle** [reply]
`meson introspect meson.build --projectinfo | jq .version` does the job. Thanks.

**Reply by eli-schwartz** [reply]
So, like I said above. :)

... but with caveats that it may be quite inefficient about it.

**Comment by xclaesse** [comment]
IMHO it would make sense to add a subcommand to meson introspect that just parse the initial project() arguments.

**Reply by rffontenelle** [reply]
Totally agree. Distro package maintainers would love it.

**Reply by eli-schwartz** [reply]
I am actually trying to think of a good reason for this, but the only one I can think of is something like the wrapdb checks where we very specifically want to compare the version in the tarball url to the version in meson.build and assert that they match.

I don't think that's very interesting to distro package maintainers though.

**Reply by rffontenelle** [reply]
I would use it for getting pkgver in VCS packages in Arch when the repository has no git tag

**Reply by eli-schwartz** [reply]
... and I guess you would end up with a revision count incrementing from the root commit, not from the latest version. Sad, but I suppose it does the job.

Seems pretty niche enough to not require a dedicated subcommand though.
